docs/architecture/adr-043-nft-module.md

```diff
@@ -0,0 +1,253 @@
+# ADR 43: NFT Module
+
+## Changelog
+
+- 05.05.2021: Initial Draft
+
+## Status
+
+DRAFT
+
+## Abstract
+
+This ADR defines a generic NFT module of `x/nft` which supports NFTs as a `proto.Any` and contains `NFT` as the default implementation.
+
+## Context
+NFTs are more digital assets than only crypto arts, which is very helpful for accruing value to ATOM. As a result, Cosmos Hub should implement NFT functions and enable a unified mechanism for storing and sending the ownership representative of NFTs as discussed in https://github.com/cosmos/cosmos-sdk/discussions/9065.
+
+As was discussed in [#9065](https://github.com/cosmos/cosmos-sdk/discussions/9065), several potential solutions can be considered:
+- irismod/nft and modules/incubator/nft
+- CW721
+- DID NFTs
+- interNFT
+
+Since NFTs functions/use cases are tightly connected with their logic, it is almost impossible to support all the NFTs' use cases in one Cosmos SDK module by defining and implementing different transaction types.
+
+Considering generic usage and compatibility of interchain protocols including IBC and Gravity Bridge, it is preferred to have a very simple NFT module design which only stores NFTs by id and owner. 
+
+This design idea can enable composability that application-specific functions should be managed by other modules on Cosmos Hub or on other Zones by importing the NFT module.
+
+For example, NFTs can be grouped into collections in a collectibles module to define the behaviors such as minting, burning, etc.
+
+The current design is based on the work done by [IRISnet team](https://github.com/irisnet/irismod/tree/master/modules/nft) and an older implementation in the [Cosmos repository](https://github.com/cosmos/modules/tree/master/incubator/nft).
+
+## Decision
+
+We will create a module `x/nft`, which provides the most basic storage, query functions for other upper-level modules to call. This module implements the OCAP security model in the [ADR 033](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-033-protobuf-inter-module-comm.md) protocol.
+
+### Types
+
+We define a general NFT model and `x/nft` module only stores NFTs by id and owner.
+
+```go
+type NFT struct{
+  string id    = 1;
+  string owner = 2;
+  Data data    = 3;
+}
+
+// Data defines a minimum funtion set consistent with `ERC721Metadata` interface.
+type Data interface {
+  Name() string // A descriptive name for a collection of NFTs
+  Symbol() string // An abbreviated name for NFTs
+  URI() string // A distinct Uniform Resource Identifier (URI) for a given asset
+}
```

Comment:
If a DID is used for the Denom, this would provide:
1. A URI for locating the root NFT (in a chain namespace and specific Cosmos module or Ethereum smart contract address)
2. A URI for dereferencing the token metadata, if this is stored in the format of a DID document (on or off-chain)
3. A metadata and control model which is compatible with DIDs and all the related family of specs and tooling (as specified for Interchain Identifiers which includes properties for linked resources, services, verification relationships, verification methods and accorded rights)
4. A portable standards-based token identifier which can be transferred across networks
5. An identifier which can be used for a wide range of DID-related applications and services, such as attestations about the token, authentication and control, delegable authorisation capabilities, etc. 
6. A very large name-space for token Denoms

Whilst these are desirable benefits, implementers could still have the option of using traditional Denoms, with Monikers, or other alternatives. 

Does the Denom really have to be a moniker? With the ERC1155 Token Multi-standard we see a pattern of moving the human-readable identifiers and related metadata relative to the token and not intrinsic. This solves the issue of localisation by allowing for variations of the token identity in respect to different languages and representations. 
It also allows for a large number of tokens to use the same on-chain string by defining a (DID) URI once, for that large number of tokens. 

Adding DID fragments could allow for multiple fungible tokens to be associated with a base NFT. For instance `DID:IXO:NFT:123#shares` for tokens that represent shares in the fractional ownership of an NFT and a separate token `DID:IXO:NFT:123#governance` that enables governance functions related to the NFT.

Tying a supply of (one or more) fungible token sets with a DID Denom in Bank to the same non-fungible token Denom in baseNFT would enable fractional shares of an NFT to be issued in a way that the non-fungible asset is directly linked to its fungible derivatives. This seems to be a core requirement for many NFT use-cases. Services for functions such as `deriveShares` could then be added through application-specific customisations of baseNFT, or delivered by other modules or external services.

The bank module now allows long strings for Denom, so this should not limit using a DID for denom. However, there could be an implementation restriction with this approach as standard Regex does not allow colons as used in DID uri's From what I understand, this can be over-ridden. But is the Regex over-write on one chain likely to break on another chain when a DID Denom gets IBC transferred (may relate to #7027)?

Reply:
@ig-shaun  thanks for the great comment. I used it in the alternative design for NFT which is derived from the comments here, please have a look here: https://github.com/cosmos/cosmos-sdk/discussions/9065#discussioncomment-873206